### PR TITLE
MF-732 - Add PostgreSQL writer

### PR DESCRIPTION
**Pull request by manuio**
Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

### CODE REVIEW

**Review by drasko — DISMISSED** [review]
LGTM

File: cmd/postgres-writer/main.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
+//
+// Copyright (c) 2018
```
Comment (anovakovic01) [inline cmd/postgres-writer/main.go]: Change these to `2019` for this service.

File: cmd/postgres-writer/main.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	envDBSSLMode     = "MF_POSTGRES_DB_SSL_MODE"
+	envDBSSLCert     = "MF_POSTGRES_DB_SSL_CERT"
+	envDBSSLKey      = "MF_POSTGRES_DB_SSL_KEY"
+	envDBSSLRootCert = "MF_POSTGRES_DB_SSL_ROOT_CERT"
+)
+
+type config struct {
+	natsURL  string
+	logLevel string
+	Port     string
```
Comment (anovakovic01) [inline cmd/postgres-writer/main.go]: Why public field?

File: cmd/postgres-writer/main.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	go startHTTPServer(cfg.Port, errs, logger)
+
+	go func() {
+		c := make(chan os.Signal)
+		signal.Notify(c, syscall.SIGINT)
+		errs <- fmt.Errorf("%s", <-c)
+	}()
+
+	err = <-errs
+	logger.Error(fmt.Sprintf("Cassandra writer service terminated: %s", err))
```
Comment (anovakovic01) [inline cmd/postgres-writer/main.go]: Change `Cassandra` to `Postgres`.

File: docker/addons/postgres-writer/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+###
+
+version: "3"
+
+networks:
+  docker_mainflux-base-net:
+    external: true
+
+volumes:
+  mainflux-postgres-volume:
```
Comment (anovakovic01) [inline docker/addons/postgres-writer/docker-compose.yml]: Maybe rename this one to `mainflux-postgres-writer-volume`?

File: writers/postgres/init.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            value         FLOAT,
+            string_value  TEXT,
+            bool_value    BOOL,
+            data_value    TEXT,
+            value_sum     FLOAT,
+            time          FlOAT,
+            update_time   FLOAT,
+            link          TEXT,
+						PRIMARY KEY (channel)
+					)`,
```
Comment (anovakovic01) [inline writers/postgres/init.go:72]: Please fix schema formatting. More importantly, `channel` cannot be message primary ID (it's not unique).

File: writers/postgres/messages.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	dbth := toDBMessage(msg)
+
+	_, err := pr.db.NamedExec(q, dbth)
+	if err != nil {
+		pqErr, ok := err.(*pq.Error)
+		if ok {
+			switch pqErr.Code.Name() {
+			case errInvalid:
+				return nil
+			}
```
Comment (anovakovic01) [inline writers/postgres/messages.go:52]: Not sure if this is the right way to handle this error.

**Review by anovakovic01 — CHANGES_REQUESTED** [review]

File: writers/postgres/messages.go (hunk 6 above)
Comment (chombium) [reply writers/postgres/messages.go:52]: That's the way to check for PostgreSQL errors. I'm not sure about ignoring the "invalid_text_representation" errors. The protobuf and the go compiler will protect us from setting wrong string representations for the parameter values, but I think one extra check does not hurt.
@manuio What's the idea with ignoring this kind of errors?

Comment (anovakovic01) [reply writers/postgres/messages.go:52]: Yes, that is what I meant, `return nil` is just ignoring it. Also, can there be any other errors from Postgres?

Comment (chombium) [reply writers/postgres/messages.go:52]: sure there can be other errors. I also think we should also "repack" the [pq.error](https://github.com/lib/pq/blob/master/error.go) error code and error message and return it as go error.

Comment (anovakovic01) [reply writers/postgres/messages.go:52]: Yes, I agree. @manuio you can check out `things` service implementation for example.

Comment (manuio) [reply writers/postgres/messages.go:52]: Good catch, I removed the error and then forgot... I'll fix it

Comment (manuio) [reply writers/postgres/messages.go]: Good catch, I removed and then forgot... I'll fix it

**Review by drasko — DISMISSED** [review]
LGTM

**Review by anovakovic01 — CHANGES_REQUESTED** [review]
Almost done.

**Review by anovakovic01 — APPROVED** [review]
LGTM!

**Review by drasko — APPROVED** [review]
LGTM

**Pull request merged**